### PR TITLE
Fixes #93: permissive parsing

### LINKED ISSUE #93: Consider parsing the policy according to the spec
We keep opening bugs on other implementations around parsing, but we don't follow the spec ourselves. 
Do not bail out on first parsing error. Reference: http://www.w3.org/TR/CSP/#policy-parsing

### PR DESCRIPTION
**Pull request by shekyan**

### CODE REVIEW

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 1]
```diff
@@ -10,82 +9,141 @@
...
+        this.error(this.getCurrentToken(), message);
+    }
+
+    private void error(@Nullable Token token, @Nonnull String message) {
+        if (this.noticesOut != null) {
+            this.noticesOut.add(this.createNotice(token, Notice.Type.ERROR, message));
+        }
+    }
+
+    private void info(@Nonnull String message) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: Are these notice functions that work on the current token used at all? If we can, we should try to drop them.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 2]
```diff
@@ -110,37 +164,38 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
         return false;
     }
 
-    @Nonnull protected ParseException createUnexpectedEOF(@Nonnull String message) {
-        return this.createError(message);
+    @Nonnull protected DirectiveValueParseException createError(@Nonnull String message) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: We should be able to remove all the `createError` methods now.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 3]
```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
+                this.error(String.format(
+                    "Expecting directive-value but found U+%04X (%s). Non-ASCII and non-printable characters must be percent-encoded",
+                    cp, new String(new int[] {cp}, 0, 1)));
+                throw INVALID_DIRECTIVE_VALUE;
+            }
+        }
+        return result;
+    }
+
+    private void enforceMissingDirectiveValue(@Nonnull String tokenValue) throws DirectiveParseException {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: Let's rename this parameter to `directiveName`.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 4]
```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
-                            } else {
-                                port = portString.equals(":*") ?
-                                    Constants.WILDCARD_PORT :
-                                    Integer.parseInt(portString.substring(1));
-                            }
-                            String host = matcher.group("host");
-                            String path = matcher.group("path");
-                            return new HostSource(scheme, host, port, path);
+    @Nonnull
+    private List<SubDirectiveValueToken> splitByWSP(@Nonnull Token token) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: Move this method up to be next to the other helpers.

Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: We should be able to make this `static`.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 5]
```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
+        List<SubDirectiveValueToken> tokens = new ArrayList<>();
+        @Nullable Location startLocation = token.startLocation;
+        if (startLocation == null) {
+            for (String s : WSP.split(token.value)) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: `WSP.split(token.value)` should be `WSP.split(trimRHSWS(token.value))`

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 6]
```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
-            this.advance();
-            sourceExpressions.add(None.INSTANCE);
-            return sourceExpressions;
-        }
-        while (this.hasNext(DirectiveValueToken.class)) {
-            sourceExpressions.add(this.parseSourceExpression());
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            if (trimRHSWS(dv.value).equalsIgnoreCase("'none'")) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: We can get rid of this exceptional condition by keeping track of the `None` values returned by `parseSourceExpression` in the loop below. I think that would be more DRY.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 7]
```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
-            this.advance();
-            ancestorSources.add(None.INSTANCE);
-            return ancestorSources;
-        }
-        while (this.hasNext(DirectiveValueToken.class)) {
-            ancestorSources.add(this.parseAncestorSource());
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            if (trimRHSWS(dv.value).equalsIgnoreCase("'none'")) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: We can do the same thing here.

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 8]
```diff
@@ -366,67 +548,154 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
...
 
-    @Nonnull private SandboxValue parseSandboxToken() throws ParseException {
-        Token token = this.advance();
-        Matcher matcher = Constants.sandboxTokenPattern.matcher(token.value);
+    @Nonnull private SandboxValue parseSandboxToken(@Nonnull SubDirectiveValueToken token) throws DirectiveValueParseException {
+        Matcher matcher = Constants.sandboxEnumeratedTokenPattern.matcher(token.value);
         if (matcher.find()) {
             return new SandboxValue(token.value);
+        } else {
+            this.warn(token, "The sandbox directive should contain only allow-forms, allow-modals, "
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: How is this not an error? And if we know all of the possible values, shouldn't `SandboxValue` be an `enum` now?

File: src/main/java/com/shapesecurity/salvation/data/Notice.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    @Override public String toString() {
+        return type.getValue() + ": " + message;
+    }
+
+    @Nonnull
+    public static ArrayList<Notice> getAllErrors(@Nullable ArrayList<Notice> notices) {
+        if (notices == null) {
+            return new ArrayList<>();
+        }
+        return notices.stream().filter(notice -> notice.type == Type.ERROR).collect(Collectors.toCollection(ArrayList::new));
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/data/Notice.java]: Add `isError`, `isWarning`, and `isInfo` functions that do `this.type == Type.ERROR`, etc. and then use `Notice::isError` here and similar below.

File: src/main/java/com/shapesecurity/salvation/data/Notice.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        this.message = message;
+        this.type = type;
+    }
+
+    @Override public String toString() {
+        return type.getValue() + ": " + message;
+    }
+
+    @Nonnull
+    public static ArrayList<Notice> getAllErrors(@Nullable ArrayList<Notice> notices) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/data/Notice.java]: These shouldn't need to be nullable.

File: src/main/java/com/shapesecurity/salvation/Parser.java (hunk 3 above)
Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: Did you mean `DirectiveValue`?

File: src/main/java/com/shapesecurity/salvation/directives/BlockAllMixedContentDirective.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+public class BlockAllMixedContentDirective extends Directive<DirectiveValue> {
+    @Nonnull private static final String NAME = "block-all-mixed-content";
+
+    public BlockAllMixedContentDirective() {
+        super(BlockAllMixedContentDirective.NAME, Collections.emptySet());
+    }
+
+    @Nonnull @Override public Directive<DirectiveValue> construct(Set<DirectiveValue> newValues) {
+        return new BlockAllMixedContentDirective();
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directives/BlockAllMixedContentDirective.java]: `return this;`?

File: src/main/java/com/shapesecurity/salvation/Parser.java (hunk 8 above)
Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: For `sandbox`, `directive-value` can be an empty value: ```directive-value = "" / token *( RWS token )```

File: src/main/java/com/shapesecurity/salvation/Parser.java (hunk 3 above)
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: No.

**Pull request merged**